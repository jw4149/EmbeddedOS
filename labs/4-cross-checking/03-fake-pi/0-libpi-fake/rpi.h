#ifndef __RPI_H__
#define __RPI_H__
// simplest way to rewrap --- just define replacements in header --- don't
// even need .c
//
// easy, but two downsides: 
//   1. we are polluting the pi program namespace with system includes.
//   2. we aren't using the raw rpi.h so can get different behavior.
#include <stdio.h>
#include <stdlib.h>

// redefine printk as fprintf using variadic macros.
#define printk(args...) fprintf(stderr, args)

static void inline clean_reboot(void) {
    printk("DONE!!!\n");
    exit(0);
}

void notmain(void);
#endif
