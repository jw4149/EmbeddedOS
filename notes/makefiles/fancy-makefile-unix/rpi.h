#include <stdio.h>

#define printk printf
#define debug printf
