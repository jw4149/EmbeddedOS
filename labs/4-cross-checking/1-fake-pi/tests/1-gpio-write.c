#include "rpi.h"
void notmain(void) {
    output("about to turn pins [0..32) on\n");
    for(unsigned pin = 0; pin < 32; pin++) {
        output("turning pin %d on\n", pin);
        gpio_write(pin,1);
    }

    output("about to turn pins [0..32) off\n");
    for(unsigned pin = 0; pin < 32; pin++) {
        output("turning pin %d off\n", pin);
        gpio_write(pin,0);
    }
}
