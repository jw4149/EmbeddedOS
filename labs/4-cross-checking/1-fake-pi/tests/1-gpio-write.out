TRACE:0: calling pi code
TRACE:1: PUT32(0x2020001c) = 0x1
TRACE:2: PUT32(0x2020001c) = 0x2
TRACE:3: PUT32(0x2020001c) = 0x4
TRACE:4: PUT32(0x2020001c) = 0x8
TRACE:5: PUT32(0x2020001c) = 0x10
TRACE:6: PUT32(0x2020001c) = 0x20
TRACE:7: PUT32(0x2020001c) = 0x40
TRACE:8: PUT32(0x2020001c) = 0x80
TRACE:9: PUT32(0x2020001c) = 0x100
TRACE:10: PUT32(0x2020001c) = 0x200
TRACE:11: PUT32(0x2020001c) = 0x400
TRACE:12: PUT32(0x2020001c) = 0x800
TRACE:13: PUT32(0x2020001c) = 0x1000
TRACE:14: PUT32(0x2020001c) = 0x2000
TRACE:15: PUT32(0x2020001c) = 0x4000
TRACE:16: PUT32(0x2020001c) = 0x8000
TRACE:17: PUT32(0x2020001c) = 0x10000
TRACE:18: PUT32(0x2020001c) = 0x20000
TRACE:19: PUT32(0x2020001c) = 0x40000
TRACE:20: PUT32(0x2020001c) = 0x80000
TRACE:21: PUT32(0x2020001c) = 0x100000
TRACE:22: PUT32(0x2020001c) = 0x200000
TRACE:23: PUT32(0x2020001c) = 0x400000
TRACE:24: PUT32(0x2020001c) = 0x800000
TRACE:25: PUT32(0x2020001c) = 0x1000000
TRACE:26: PUT32(0x2020001c) = 0x2000000
TRACE:27: PUT32(0x2020001c) = 0x4000000
TRACE:28: PUT32(0x2020001c) = 0x8000000
TRACE:29: PUT32(0x2020001c) = 0x10000000
TRACE:30: PUT32(0x2020001c) = 0x20000000
TRACE:31: PUT32(0x2020001c) = 0x40000000
TRACE:32: PUT32(0x2020001c) = 0x80000000
TRACE:33: PUT32(0x20200028) = 0x1
TRACE:34: PUT32(0x20200028) = 0x2
TRACE:35: PUT32(0x20200028) = 0x4
TRACE:36: PUT32(0x20200028) = 0x8
TRACE:37: PUT32(0x20200028) = 0x10
TRACE:38: PUT32(0x20200028) = 0x20
TRACE:39: PUT32(0x20200028) = 0x40
TRACE:40: PUT32(0x20200028) = 0x80
TRACE:41: PUT32(0x20200028) = 0x100
TRACE:42: PUT32(0x20200028) = 0x200
TRACE:43: PUT32(0x20200028) = 0x400
TRACE:44: PUT32(0x20200028) = 0x800
TRACE:45: PUT32(0x20200028) = 0x1000
TRACE:46: PUT32(0x20200028) = 0x2000
TRACE:47: PUT32(0x20200028) = 0x4000
TRACE:48: PUT32(0x20200028) = 0x8000
TRACE:49: PUT32(0x20200028) = 0x10000
TRACE:50: PUT32(0x20200028) = 0x20000
TRACE:51: PUT32(0x20200028) = 0x40000
TRACE:52: PUT32(0x20200028) = 0x80000
TRACE:53: PUT32(0x20200028) = 0x100000
TRACE:54: PUT32(0x20200028) = 0x200000
TRACE:55: PUT32(0x20200028) = 0x400000
TRACE:56: PUT32(0x20200028) = 0x800000
TRACE:57: PUT32(0x20200028) = 0x1000000
TRACE:58: PUT32(0x20200028) = 0x2000000
TRACE:59: PUT32(0x20200028) = 0x4000000
TRACE:60: PUT32(0x20200028) = 0x8000000
TRACE:61: PUT32(0x20200028) = 0x10000000
TRACE:62: PUT32(0x20200028) = 0x20000000
TRACE:63: PUT32(0x20200028) = 0x40000000
TRACE:64: PUT32(0x20200028) = 0x80000000
TRACE: pi exited cleanly: 7 calls to random
